### ISSUE TITLE
Support for 32-bit fixed point types - fixed, fixed2, fixed3, fixed4

### ISSUE THREAD

**Issue by james7132** [Unity-Technologies/Unity.Mathematics#88]
I know that Burst is promising cross-platform floating point determinism in the future, but that's not only a high bar to cross (it's a problem almost nobody has solved before), it also limits the usability of deterministic computation to Burst compiled jobs.

Having general access to fixed point computations, be it 32 (Q16) or 64 (Q32) bit, could make it much easier to have universally available deterministic decimal computations.

**Comment by unpacklo** [comment]
I agree that having a fixed point format could be useful, the main question is which format to actually support and ensuring it is performant.  Do you have specific use cases in mind for the fixed point formats?

**Comment by nxrighthere** [comment]
My colleagues are extensively using Q48.16 for deterministic multiplayer games, they have tried several formats, and this one has a really good balance of performance and precision. For example, Q32.32 is about 2-3X slower at multiplication/division because of the implementation constraints.

**Comment by nxrighthere** [comment]
My advice is to go with Q48.16 or make it switchable between Q48.16 and Q32.32 where one can be used for fast multiplication/division, and another for a higher precision where it's required. In most cases, Q48.16 gives enough of the depth.

**Comment by unpacklo** [comment]
Is it necessary for you to have the 48 bits of range in the whole portion?

**Comment by nxrighthere** [comment]
Not really, it's only used because of performance reasons. I asked my colleagues for the actual code example:

Q48.16
```csharp
public static FP operator *(FP a, FP b) {
    a.RawValue = (a.RawValue * b.RawValue) >> FPLut.PRECISION;

    return a;
}
```

Q32.32
```csharp
public static FP operator *(FP a, FP b) {
	var al = a.m_rawValue;
	var bl = b.m_rawValue;

	var alo = (ulong)(al & 0x00000000FFFFFFFF);
	var ahi = al >> FRACTIONAL_PLACES;
	var blo = (ulong)(bl & 0x00000000FFFFFFFF);
	var bhi = bl >> FRACTIONAL_PLACES;

	var lolo = alo * blo;
	var lohi = (long)alo * bhi;
	var hilo = ahi * (long)blo;
	var hihi = ahi * bhi;

	var loResult = lolo >> FRACTIONAL_PLACES;
	var midResult1 = lohi;
	var midResult2 = hilo;
	var hiResult = hihi << FRACTIONAL_PLACES;

	bool overflow = false;
	var sum = AddOverflowHelper((long)loResult, midResult1, ref overflow);

	sum = AddOverflowHelper(sum, midResult2, ref overflow);
	sum = AddOverflowHelper(sum, hiResult, ref overflow);

	bool opSignsEqual = ((al ^ bl) & MIN_VALUE) == 0;

	if (opSignsEqual) {
		if (sum < 0 || (overflow && al > 0))
			return MaxValue;
	} else {
		if (sum > 0)
			return MinValue;
	}

	var topCarry = hihi >> FRACTIONAL_PLACES;

	if (topCarry != 0 && topCarry != -1) {
		return opSignsEqual ? MaxValue : MinValue;
	}

	if (!opSignsEqual) {
		long posOp, negOp;

		if (al > bl) {
			posOp = al;
			negOp = bl;
		} else {
			posOp = bl;
			negOp = al;
		}

		if (sum > negOp && negOp < -ONE && posOp > ONE)
			return MinValue;
	}

	return new FP(sum);
}
```

**Comment by nxrighthere** [comment]
So on practice with Q48.16, you rarely need 48 bits of range rather than using the top 32 bits as a buffer for multiplication/division. Q32.32 gives you more accuracy on decimals, but practically it doesn't carry much benefits and costs performance.

**Comment by nxrighthere** [comment]
Q48.16 is, in general, the best compromise in comparison to other formats due to LUT sizes, performance. accuracy, and so on.

**Comment by nxrighthere** [comment]
Note that the actual usable space in this case is -Q15.16 to +Q15.16.

**Comment by james7132** [comment]
Following up on this, @nxrighthere, is there currently an open source version for this Q48.16 implementation you're talking about? FixedMath.NET seems to use a Q32.32 implementation.

**Comment by nxrighthere** [comment]
@james7132 The one that I'm talking about is an in-house proprietary library. I doubt that you will find an open-source C# implementation of Q48.16, this format is about performance primarily, those who need fixed-point math are don't care about it usually and they use what commonly used.

**Comment by unpacklo** [comment]
@james7132 I understand, thanks.

When you say that it would be nice to have it be universally available, you're referring to being able to just start any Unity project and being assured that fixed point types are ready to use? My expectation is that fixed point types would be implemented by companies that require them and those libs would be used internally on their own projects.

**Comment by nxrighthere** [comment]
@unpacklo I think it's more about that you have chosen a suboptimal path instead of using the best available tool for the job. Enforced deterministic cross-platform floating-points through compilation means a lot of tradeoffs, it most likely means IEEE 754 compliance, which means trading performance and accuracy by disabling floating-point optimizations. So you might end up with a situation where a particular fixed-point format based on integers will be faster, just because of that.

The second problem is the engineering effort. This feature is in development for more than two years already.

**Comment by nxrighthere** [comment]
And my personal design preference (as a network programmer): when you work with abstracted types like `FP` instead of `float` with enforced (per method?) compilation, you have a type-safe, explicit facility to distinguish the two to not turn it into an unmaintainable mess with deterministic and non-deterministic floats. But if this compilation option will be available per project basis instead of per method, then you effectively killing performance where it shouldn't.

**Comment by james7132** [comment]
@unpacklo I mean that in the sense that determinism isn't bound to a specific compiler. I understand that Unity.Mathematics is best used in tandem with Burst, but that should not be it's only priority. Burst, AFAIK, is only bound for C# jobs, and cannot be easily utilized outside of those contexts due to strict invariants about it's core assumptions. If this has changed, please correct my understanding. The use of fixed point computation is to have deterministic non-integer computations regardless of what compiler, execution context, etc. As @nxrighthere, it's difficult to tell when a float operation is deterministic since the state of determinism is bound to the compiler, not a specific datatype. If I pull out the results of a C# job compiled with Burst then run even one non-deterministic operation on it from the main thread, all the effort in keeping it deterministic is lost, whereas with fixed-point I can strictly guarantee that no matter what operation I am doing on it, regardless of CPU architecture or compiler, the output is exactly the same.

**Comment by unpacklo** [comment]
@james7132 thanks for the clarification, I understand now.

I created an issue in our internal issue tracker so we can evaluate this feature!

**Comment by IronWarrior** [comment]
Reviving this a bit, [this repo is an example of 48.16 FP in C#](https://github.com/RomanZhu/FixedPoint-Sharp). I have a [small library that provides Unity integration](https://github.com/IronWarrior/FixedPoint-Sharp-Unity) (property drawers etc). An example of a game engine using 48.16 is [Photon Quantum](https://www.photonengine.com/quantum) (for deterministic networking), which has a 2D and 3D physics engine.
